text/0000-let-else.md

```diff
@@ -0,0 +1,492 @@
+and also that the resulting `match` code includes syntax boilerplate (e.g. the block) that could theoretically be reduced today but also interferes with rustfmt's rules:
+
+```rust
+let features = match geojson {
+    GeoJson::FeatureCollection(features) => features,
+    _ => {
+        return Err(format_err_status!(
+            422,
+            "GeoJSON was not a Feature Collection",
+        ));
+    }
+};
+```
+
+However, with if-let this could be more succinct & clear:
+
+```rust
+let GeoJson::FeatureCollection(features) = geojson else {
+    return Err(format_err_status!(
+        422,
+        "GeoJSON was not a Feature Collection",
+    ));
+};
+```
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+A common pattern in non-trivial code where static guarantees can not be fully met (e.g. I/O, network or otherwise) is to check error cases when possible before proceeding,
+and "return early", by constructing an error `Result` or an empty `Option`, and returning it before the "happy path" code.
+
+This pattern serves no practical purpose to a computer, but it is helpful for humans interacting with the code.
+Returning early helps improve code clarity in two ways:
+- Ensuring the returned result in near the conditional, visually, as the following logic may be lengthy.
+- Reduces rightward shift, as the error return is now in the block, rather than the following logic.
+
+This RFC proposes _(Rust provides)_ an extension to `let` assignment statements to help with this pattern, an `else { }` which can follow a pattern match
+as a `let` assigning statement:
+
+```rust
+let Some(a) = an_option else {
+    // Called if `an_option` is not `Option::Some(T)`.
+    // This block must diverge (stop executing the existing context to the parent block or function).
+    return;
+};
+
+// `a` is now in scope and is the type which the `Option` contained.
+```
+
+This is a counterpart to `if let` expressions, and the pattern matching works identically, except that the value from the pattern match
+is assigned to the surrounding scope rather than the block's scope.
+
+# Reference-level explanations
+[reference-level-explanation]: #reference-level-explanation
+
+let-else is syntactical sugar for `match` where the non-matched case diverges.
+```rust
+let pattern = expr else {
+    /* diverging expr */
+};
+```
+desugars to
+```rust
+let (each, binding) = match expr { 
+    pattern => (each, binding),
+    _ => { 
+        /* diverging expr */
+    }
+};
+```
+
+Any expression may be put into the expression position except an `if {} else {}` as explain below in [drawbacks][].
+While `if {} else {}` is technically feasible this RFC proposes it be disallowed for programmer clarity to avoid an `... else {} else {}` situation.
+Rust already provides us with such a restriction, [`ExpressionWithoutBlock`][expressions].
```

Comment:
Nothing else in the language right now uses this terminal to prevent `ExpressionWithBlock` from appearing in certain contexts. It is used more as way to allow description of situations where fewer terminals are required after `ExpressionWithBlock`s used in certain contexts. For example:

```
MatchArms :
   ( MatchArm => ( ExpressionWithoutBlock , | ExpressionWithBlock ,? ) )*
   MatchArm => Expression ,?
```

This kind of rule still allows for use of e.g. generic `expr` metavariable in macros-by-example to work.

As described right now, in macros by-example you would be required to introduce additional structure (`GroupedExpression`?) in order to be able to use the matched `expr` structures in let-else statements, which seems difficult to discover and unintuitive.

Could this section of the RFC describe this feature in terms of changes/additions to the grammar, rather than prose; or both?

Reply:
The suggestion to handle expressions with a block is indeed to use a `GroupedExpression` like `(unsafe { })` or `(if { } else { })`. I can add that note.

> Could this section of the RFC describe this feature in terms of changes/additions to the grammar, rather than prose; or both?

I'll need some on help on that; I'm not super familiar with the grammar at this point!

The suggestion to restrict to `ExpressionWithoutBlock` came from [a recent comment on the old RFC](https://github.com/rust-lang/rfcs/pull/1303#issuecomment-840744369) by @bstrie.